dwh_design_conventions.md

```diff
@@ -0,0 +1,177 @@
+tangible warehouse.
+
+This can be as simple as making all intermediate scripts
+a mix of ephemeral models (compiled SQL code that dbt injects in place of `ref()`
+statements) and tables in an unexposed schema. It can be as fancy as setting
+up Raw and Analytics logical databases in Snowflake, with a `salesforce.account`
+in each.
+
+### Staging Raw Data
+
+The staging model is the atomic unit of data modeling. Each model bears a 
+one-to-one relationship with the raw data table it represents: 
+`salesforce.account` becomes `stg_salesforce_account`. It has the same
+granularity, but the columns have been renamed, recast, or usefully
+reconsidered.
+
+Every raw data table should still flow through a base model of the form:
+```sql
+with source as (
+    
+    select * from raw_schema.raw_table
+    
+),
+
+renamed as (
+    
+    select
+    
+        firstname as first_name,
+        createdat::date as created_at,
+        ...
+    
+)
+
+select * from renamed
+```
+Depending on the quality of raw data, this base model could also be a staging
+model. A staged view might just as well warrant several models' worth of cleaning, 
+correcting, and categorizing. Raw data comes as it is; staging data is clean, 
+presentable, and ready for the curtain to rise.
+
+Staging models **can** have joins in them to field additional columns for context 
+or enrichment; add rows through unions and remove them through filters;
+deduplicate a natural key or hash together a 
+[surrogate one](https://github.com/fishtown-analytics/dbt-utils#surrogate_key-source); 
+but they should **not** have `group by` aggregations that would change their 
+granularity. The `stg_salesforce_account` model should have one account per row, 
+and `account_id` as its primary key.
+
+In our experience, it is most helpful to have all of these models within a
+single `models/staging` folder, with one subfolder for each datasource:
+
+```yml
+staging:
+    salesforce:
+        schema: stg_salesforce
+        base:
+            materialized: ephemeral
+            # salesforce_account_base.sql
+        stg:
+            materialized: view
+            # stg_salesforce_account.sql
+    snowplow:
+        schema: stg_snowplow
+        ...
+```
+
+### Marts of Facts
+
+Complex logical rollups can and should happen across several models, with
+discrete steps and abstractions split into queries of 100 lines or fewer. The
+end of these rollups is a table, exposed to end-users and aptly prefixed `fct_`:
+it contains the facts that a good reporter seeks.
+
+Marts are stores of one or more fact tables, related primarily by their use case
+in analysis and reporting. They could be organized by business unit, high-level 
+concept, or intended set of end-users: `marketing`, `finance`, `product`, 
+etc. Some fact tables form the basis for other ones and are used across the
+project; you might gird these keystones into a `core` folder, and `ref()` them
+at will.

```

Comment:
In traditional Kimball-style warehousing, models like these are filled with so-called "conformed dimensions".

https://www.kimballgroup.com/data-warehouse-business-intelligence-resources/kimball-techniques/dimensional-modeling-techniques/conformed-dimension/